File: frontend/lib/server/grounding.ts
/**
 * grounding.ts — Integrated 3-Stage Grounding & Semantic Engine.
 *
 * Implements the methodology from the paper §2.3, §3.1.1, §3.1.2, §3.1.3:
 * 1. Layer A (Step 3 Grounding): Domain-aware Lexical Matcher with CUSTOM_WORDS.
 * 2. Layer B (Step 1 & 2): Gemini-assisted Stopword Filtering & Taxonomy Classification.
 * 3. Layer C: UI Review & Selection Gate.
 */

import {
  CUSTOM_WORDS,
  GENERIC_TERMS_PREDEFINED,
  domainAwareTokenize,
  filterStopwordsWithGemini,
  normalizeThaiText,
} from "./semantic-pipeline";
import {
  CANONICAL_TAXONOMY_PATHS,
  classifyNewKeywordsWithGemini,
} from "./taxonomy-classifier";
import type { KeywordProposal } from "@/lib/types";

const STOPWORD_PATHS = new Set(["stopword", "sw"]);
const JACCARD_MIN = 0.34;

export interface GroundingVocabEntry {
  id: number;
  name: string;
  taxonomyPath: string;
}

/**
 * Layer A: Rule-based exact_token matcher (Paper §3.1.3)
 */
export function autoGroundKeywords(
  item: { name?: string; description?: string; category_group?: string; performance_type?: string },
  vocab: GroundingVocabEntry[],
): number[] {
  if (!vocab || vocab.length === 0) return [];
  const textRaw = [item.name, item.description, item.category_group, item.performance_type]
    .map((value) => value ?? "")
    .join(" ");
  const text = normalizeThaiText(textRaw);
  if (!text) return [];

  const vocabNames = vocab.map((v) => v.name);
  const itemTokens = domainAwareTokenize(text, vocabNames);

  const seen = new Set<number>();
  const matched: number[] = [];

  for (const entry of vocab) {
    if (STOPWORD_PATHS.has(entry.taxonomyPath.toLocaleLowerCase())) continue;
    if (seen.has(entry.id)) continue;
    if (GENERIC_TERMS_PREDEFINED.has(entry.name.trim())) continue;

    const kwNorm = normalizeThaiText(entry.name);
    if (!kwNorm || kwNorm.length < 2) continue;
    const kwTokens = domainAwareTokenize(entry.name, [entry.name]);

    // 1. Exact token match
    if (itemTokens.has(kwNorm)) {
      seen.add(entry.id);
      matched.push(entry.id);
      continue;
    }
    // 2. Substring match
    if (text.includes(kwNorm)) {
      seen.add(entry.id);
      matched.push(entry.id);
      continue;
    }
    // 3. Token-set Jaccard overlap
    if (kwTokens.size > 0 && itemTokens.size > 0) {
      let inter = 0;
      for (const token of kwTokens) if (itemTokens.has(token)) inter += 1;
      const union = kwTokens.size + itemTokens.size - inter;
      if (union > 0 && inter / union >= JACCARD_MIN) {
        seen.add(entry.id);
        matched.push(entry.id);
      }
    }
  }
  return matched;
}

/**
 * Extract candidate phrase fragments from text that are not yet matched in the vocabulary.
 */
function extractUnmatchedCandidatePhrases(
  text: string,
  matchedVocabNames: Set<string>,
): string[] {
  const normText = normalizeThaiText(text);
  if (!normText) return [];

  // Match words from CUSTOM_WORDS and potential domain phrases
  const candidates = new Set<string>();

  for (const cw of CUSTOM_WORDS) {
    const normCw = normalizeThaiText(cw);
    if (normCw && normText.includes(normCw) && !matchedVocabNames.has(normCw)) {
      if (!GENERIC_TERMS_PREDEFINED.has(normCw)) {
        candidates.add(cw);
      }
    }
  }

  // Also extract space-delimited segments of 3-30 chars that look like domain terms
  const rawSegments = text.split(/[\s,.;:()\[\]{}'"\-–—\/]+/);
  for (const seg of rawSegments) {
    const s = seg.trim();
    if (s.length >= 3 && s.length <= 30 && !GENERIC_TERMS_PREDEFINED.has(s)) {
      const normS = normalizeThaiText(s);
      if (!matchedVocabNames.has(normS) && !candidates.has(s)) {
        candidates.add(s);
      }
    }
  }

  return Array.from(candidates).slice(0, 40); // Cap batch size for Gemini
}

/**
 * End-to-end Semantic Pipeline Grounding (Step 1 -> Step 2 -> Step 3)
 */
export async function executeSemanticPipelineGrounding(
  item: { name: string; description?: string; category_group?: string; performance_type?: string },
  vocab: GroundingVocabEntry[],
): Promise<KeywordProposal[]> {
  const proposals: KeywordProposal[] = [];
  const matchedVocabNames = new Set<string>();

  // 1. Layer A (Grounding from Master 584 Vocab)
  const layerAIds = autoGroundKeywords(item, vocab);
  for (const id of layerAIds) {
    const entry = vocab.find((v) => v.id === id);
    if (entry) {
      matchedVocabNames.add(normalizeThaiText(entry.name));
      proposals.push({
        id: entry.id,
        name: entry.name,
        source: "auto",
        confidence: 1.0,
        taxonomy_path: entry.taxonomyPath,
        is_new: false,
      });
    }
  }

  // 2. Step 1 (Stopword Filtering) & Step 2 (Taxonomy Classification) for Candidate Words
  const fullText = `${item.name} ${item.description || ""} ${item.category_group || ""} ${item.performance_type || ""}`;
  const candidatePhrases = extractUnmatchedCandidatePhrases(fullText, matchedVocabNames);

  if (candidatePhrases.length > 0 && process.env.GEMINI_API_KEY?.trim()) {
    try {
      // Step 1: Filter Stopwords
      const nonStopwords = await filterStopwordsWithGemini(candidatePhrases, item);
      // Remove any that happen to match existing proposals
      const newNonStopwords = nonStopwords.filter(
        (w) => !matchedVocabNames.has(normalizeThaiText(w)),
      );

      if (newNonStopwords.length > 0) {
        // Step 2: Classify into 23 Taxonomy Paths
        const classifications = await classifyNewKeywordsWithGemini(newNonStopwords);
        let tempId = -1;
        for (const [word, cls] of classifications.entries()) {
          proposals.push({
            id: tempId--,
            name: word,
            source: "llm",
            confidence: cls.confidence,
            taxonomy_path: cls.fullPath,
            is_new: true,
          });
        }
      }
    } catch (e) {
      console.warn("[grounding] Semantic pipeline AI step encountered error:", e);
    }
  }

  return proposals;
}

File: backend/app/services/grounding.py
"""
services/grounding.py — Layer A + Layer B + Layer C keyword grounding.

Mirrors the methodology in the thesis paper §2.3 + §3.1.3:

* **Layer A** — rule-based grounding: exact token, substring, and
  token-set Jaccard matching between item text and the existing
  vocabulary. Conservative (paper F1 = 0.886).
* **Layer B** — Gemini-assisted grounding: ask the LLM to suggest at
  most ``max_suggest`` keywords from the existing vocabulary that best
  describe the item. Optional; falls back to ``[]`` when the API key is
  missing or ``grounding_use_llm=False``.
* **Layer C** — handled by the admin UI: the user edits the proposed
  keyword list before committing. The backend exposes
  ``KeywordProposal`` + ``ItemCommit`` schemas for that flow.

The vocabulary is built lazily from the in-memory artifact's catalog
via ``load_vocab`` and cached as a module singleton.
"""
from __future__ import annotations

import json
import logging
import re
import unicodedata
from typing import Dict, List, NamedTuple, Optional, Set

from ..core.config import get_settings
from ..model_loader import ArtifactLoader, get_singleton


logger = logging.getLogger("recsys.grounding")


# --- Vocab entry ------------------------------------------------------------


class VocabEntry(NamedTuple):
    """A single keyword in the universe of selectable vocabulary."""

    id: int
    name: str
    taxonomy_path: str = ""


# --- Vocab cache ------------------------------------------------------------


_vocab_cache: Optional[List[VocabEntry]] = None
_vocab_by_name: Dict[str, VocabEntry] = {}


# taxonomy_paths the paper explicitly excludes (stopword category).
_STOPWORD_PATHS = {"stopword", "sw"}


def load_vocab(loader: ArtifactLoader) -> List[VocabEntry]:
    """Build the deduplicated keyword universe from the artifact.

    Reads ``loader.items["keyword_names"]`` and the parallel
    ``taxonomy_paths`` column. Cached so repeated ingest calls share
    the same VocabEntry objects.

    Synthetic corpora may have non-integer ids — we only emit ids that
    fit in ``int`` (int64). Entries with no id column are skipped.
    """
    global _vocab_cache
    if _vocab_cache is not None:
        return _vocab_cache
    if loader is None or loader.items is None:
        return []

    items = loader.items
    has_kw = "keyword_names" in items.columns
    has_tax = "taxonomy_paths" in items.columns
    if not has_kw:
        _vocab_cache = []
        return _vocab_cache

    entries: Dict[str, VocabEntry] = {}
    # The artifact assigns synthetic ids when running with --synthetic-embeddings;
    # we derive a stable id per name via stable_id("keyword", name) so the
    # proposals survive across builds.
    from ._ids import stable_id

    for idx in range(len(items)):
        kws = items.iloc[idx].get("keyword_names") or []
        taxs = items.iloc[idx].get("taxonomy_paths") if has_tax else None
        if not isinstance(kws, (list, tuple)):
            continue
        if not taxs or not isinstance(taxs, (list, tuple)):
            taxs = [""] * len(kws)
        for kw_name, tax_path in zip(kws, taxs):
            if not isinstance(kw_name, str) or not kw_name.strip():
                continue
            name = kw_name.strip()
            tax = (tax_path or "").strip()
            if tax.lower() in _STOPWORD_PATHS:
                continue
            key = name.lower()
            if key in entries:
                continue
            entries[key] = VocabEntry(
                id=stable_id("keyword", name),
                name=name,
                taxonomy_path=tax,
            )
    _vocab_cache = list(entries.values())
    _vocab_by_name.clear()
    _vocab_by_name.update({v.name.lower(): v for v in _vocab_cache})
    return _vocab_cache


def reset_vocab_cache() -> None:
    """Drop the vocab cache. Test helper for ``ArtifactLoader`` reload."""
    global _vocab_cache
    _vocab_cache = None
    _vocab_by_name.clear()


# --- Tokenisation -----------------------------------------------------------


def _normalize(text: str) -> str:
    if not isinstance(text, str):
        return ""
    text = unicodedata.normalize("NFKC", text).strip().lower()
    # Strip Thai diacritics (่-์) and zero-width chars.
    text = re.sub(r"[่-์​-‏]", "", text)
    return text


def _tokenize_thai(text: str) -> Set[str]:
    """Tokenise Thai text into a set of normalised tokens.

    Uses ``pythainlp.word_tokenize(..., engine="newmm")`` when available;
    falls back to whitespace split if pythainlp isn't installed (dev
    sandbox without network).
    """
    text = _normalize(text)
    if not text:
        return set()
    try:  # pragma: no cover — exercised when pythainlp is installed
        from pythainlp.tokenize import word_tokenize  # type: ignore

        return {tok for tok in word_tokenize(text, engine="newmm") if tok}
    except Exception:
        return {tok for tok in text.split() if tok}


# --- Layer A ----------------------------------------------------------------


# Minimum Jaccard threshold (paper's "high-precision lexical evidence" — at
# 95 threshold only exact matches and token-set matches pass; we use 0.34
# for token-set to capture small-vocab items).
_JACCARD_MIN = 0.34


def auto_ground_keywords(
    item: Dict,
    vocab: List[VocabEntry],
    *,
    jaccard_min: float = _JACCARD_MIN,
) -> List[int]:
    """Layer A: rule-based auto-ground. Returns deduplicated vocab ids.

    Three matchers run, in order of strictness:

    1. **Exact token** — vocab.name appears as a token in item name+description.
    2. **Substring** — vocab.name is a substring of item text (catches
       multi-word compound keywords like "ชุดไทย" or "การแสดงนาฏศิลป์").
    3. **Token-set Jaccard** — ``|kw_tokens ∩ item_tokens| /
       |kw_tokens ∪ item_tokens| >= jaccard_min``.

    Stopword entries (taxonomy_path in ``_STOPWORD_PATHS``) are skipped.
    """
    if not vocab or not isinstance(item, dict):
        return []
    text_raw = " ".join(
        str(item.get(k, "") or "")
        for k in ("name", "description", "category_group", "performance_type")
    )
    text = _normalize(text_raw)
    if not text:
        return []
    item_tokens = _tokenize_thai(text)

    seen: Set[int] = set()
    matched: List[int] = []

    for entry in vocab:
        if entry.taxonomy_path.lower() in _STOPWORD_PATHS:
            continue
        if entry.id in seen:
            continue
        kw_norm = _normalize(entry.name)
        if not kw_norm:
            continue
        kw_tokens = _tokenize_thai(entry.name)

        # 1. Exact token match.
        if kw_norm in item_tokens:
            seen.add(entry.id)
            matched.append(entry.id)
            continue
        # 2. Substring match (covers multi-word keywords).
        if kw_norm in text:
            seen.add(entry.id)
            matched.append(entry.id)
            continue
        # 3. Token-set Jaccard.
        if kw_tokens and item_tokens:
            inter = len(kw_tokens & item_tokens)
            union = len(kw_tokens | item_tokens)
            if union and (inter / union) >= jaccard_min:
                seen.add(entry.id)
                matched.append(entry.id)
    return matched


# --- Layer B ----------------------------------------------------------------


def _call_gemini(prompt: str, model_name: str, api_key: str) -> str:
    """Call Gemini and return the raw response text.

    Imported lazily so the rest of the module loads without the dependency.
    """
    import google.generativeai as genai  # type: ignore

    genai.configure(api_key=api_key)
    model = genai.GenerativeModel(model_name)
    resp = model.generate_content(
        prompt,
        generation_config={
            "temperature": 0.0,
            "response_mime_type": "application/json",
        },
    )
    return resp.text or ""


_GEMINI_PROMPT = """You are an expert curator of Thai performing-arts metadata.

Given an item and the controlled vocabulary, return at most {max_suggest} vocabulary
names that culturally and contextually describe the item. Only return names from
the vocabulary — do not invent new keywords. If nothing matches, return an empty
array.

Item name: {name}
Item description: {description}
Category group: {category}
Performance type: {ptype}

Vocabulary (name — taxonomy path):
{vocab}

Respond strictly as JSON in this shape:
{{"names": ["<vocab name>", ...]}}
"""


def llm_ground_keywords(
    item: Dict,
    vocab: List[VocabEntry],
    *,
    max_suggest: int = 8,
) -> List[int]:
    """Layer B: ask Gemini for keyword suggestions over the existing vocab.

    Returns ``[]`` if the API key is missing or
    ``settings.grounding_use_llm=False``.
    """
    settings = get_settings()
    if not settings.grounding_use_llm or not settings.gemini_api_key:
        return []
    if not vocab or not isinstance(item, dict):
        return []

    # Cap the vocab we ship to Gemini — the catalog has hundreds of keywords.
    sample = vocab[:200]
    vocab_block = "\n".join(
        f"- {v.name} — {v.taxonomy_path or '(no path)'}" for v in sample
    )
    prompt = _GEMINI_PROMPT.format(
        max_suggest=max_suggest,
        name=item.get("name", "") or "",
        description=item.get("description", "") or "",
        category=item.get("category_group", "") or "",
        ptype=item.get("performance_type", "") or "",
        vocab=vocab_block,
    )
    try:
        raw = _call_gemini(prompt, settings.gemini_model, settings.gemini_api_key)
    except Exception as exc:  # noqa: BLE001
        logger.warning("Gemini grounding failed: %s", exc)
        return []
    try:
        parsed = json.loads(raw)
        names = parsed.get("names", []) if isinstance(parsed, dict) else []
    except (ValueError, TypeError):
        logger.warning("Gemini grounding returned non-JSON: %r", raw[:200])
        return []
    if not isinstance(names, list):
        return []

    matched: List[int] = []
    seen: Set[int] = set()
    for n in names:
        if not isinstance(n, str):
            continue
        entry = _vocab_by_name.get(_normalize(n)) or _vocab_by_name.get(n.strip().lower())
        if entry is None or entry.id in seen:
            continue
        seen.add(entry.id)
        matched.append(entry.id)
        if len(matched) >= max_suggest:
            break
    return matched


# --- Combined entry point ---------------------------------------------------


def ground_keywords(
    item: Dict,
    vocab: List[VocabEntry],
    *,
    use_llm: bool = True,
    jaccard_min: float = _JACCARD_MIN,
    max_llm: int = 8,
) -> Dict[str, List[int]]:
    """Run Layer A then Layer B; return per-layer ids plus the merged set.

    ``merged_ids`` preserves Layer A first (stable, deterministic) followed
    by Layer B additions.
    """
    layer_a_ids = auto_ground_keywords(item, vocab, jaccard_min=jaccard_min)
    layer_b_ids = llm_ground_keywords(item, vocab, max_suggest=max_llm) if use_llm else []
    seen: Set[int] = set(layer_a_ids)
    merged = list(layer_a_ids)
    for kid in layer_b_ids:
        if kid in seen:
            continue
        seen.add(kid)
        merged.append(kid)
    return {
        "layer_a_ids": layer_a_ids,
        "layer_b_ids": layer_b_ids,
        "merged_ids": merged,
    }


def vocab_for_loader(loader: Optional[ArtifactLoader] = None) -> List[VocabEntry]:
    """Convenience: load vocab from the singleton loader if not provided."""
    if loader is None:
        loader = get_singleton()
    return load_vocab(loader)